# 01 Importing libraries


In [11]:
# import libraries 

In [12]:
import pandas as pd
import numpy as np
import os

# 02 Importing 'orders' and 'products' csv dataset 

In [13]:
#creating a path string
path =r"C:\Users\Lemlem TH\Desktop\CareerFoundry\Course\Data Immersion\Achievement 4\Instacart Basket Analysis"

In [14]:
#importing the 'products' csv dataset 
df_prods=pd.read_csv(os.path.join(path,'02 Data', 'Original Data','products.csv' ))



In [15]:
#importing the 'orders_wrangled.csv' dataset 
df_ords=pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'))

# 03 Mixed type data

In [16]:
# Create a test dataframe 
df_test=pd.DataFrame()

In [17]:
#Create a mixed type column
df_test['mix']=['a','b',1, True]

In [18]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [19]:
# check if the df_test dataframe conatins  any mixed-type columns


for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [20]:
# convert the data type of the 'mix' column to str
df_test['mix']=df_test['mix'].astype('str')

# 04 Missing values

## Finding missing values

In [21]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [22]:
# view the observation with the missingvalue in the product:name column
df_nan=df_prods[df_prods['product_name'].isnull()==True] # creating a sub-set dataframe containing only observation with missing value 

In [23]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## Addressing missing values

In [24]:
# Imputing using the mean can be done using: First finding mean using df.describe() then run this function: df['column with missings'].fillna(mean value, inplace=True)
# Imputing using the median can be done using: First finding the median using df.median() then run this function: df['column with missings'].fillna(median value, inplace=True)


In [25]:
# Filtering out observation with missing value (creating a new dataframe only with non-missing value)
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]


In [26]:
# Dropping all missing values 
df_prods.dropna(inplace=True)

In [27]:
# Dropping missing values from a particular column would be : df_prods.dropna(subset = [‘product_name’], inplace = True)


## Duplicates

In [28]:
# Finding duplicates 
df_dups= df_prods_clean[df_prods_clean.duplicated()]


In [29]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [30]:
# creat new dataframe that doesnt include  duplicates
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [31]:
# checking the number of rows for the new dataframe
df_prods_clean_no_dups.shape

(49672, 5)

In [32]:
#  Exporting the new 'df_prods_clean_no_dups' dataframe to the 'Prepared Data' folder as 'products_checked.csv'

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task

In [33]:
df_ords.shape

(3421083, 7)

In [34]:
# Step 2 Run the df.describe() function on df_ords dataframe
df_ords.describe()


Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  hour_at_time_of_order  days_since_prior_order  
count        3.421083e+06           3.421083e+06            3.214874e+06  
mean         2.776219e+00           1.345202e+01            1.111484e+01  
std          2.046829e+00           4.226088e+00            9.206737e+00  
min          0.000000e+00           0.000000e+00            0.000000e+00  
25%          1.000000e+00           1.000000e+01            4.000000e+00  
50%          3.000000e+00           1.300000e+01            7.000000e+00  
75%          5.000000e+00           1.600000e+01            1.500000e+01  
max          6.000000e+00           2.300000e+01            3.000000e+01

The minimum value of orders_day_of_week variable cant be 0. This is assuming that the column refers to the 7 days of a week. The minimum can be 1 and the maximum 7.


In [35]:
# Step 3 Check for mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [36]:
# Step 4 If you find mixed type data, fix it.


There is no mixed-type data in the df_ords dataframe.

In [37]:
# Step 5 Run a check for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_at_time_of_order          0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values for the days_since_prior_order variable. Other columns dont have any missing values. The missing value in the days_since_prior_order column might be for clinets that have an order on the current day (the day the data was collected)

In [38]:
#Step 6 Address the missing values using an appropriate method 
df_ords['days_since_prior_order'].fillna(0, inplace=True)

I chose to replace the missing value with a value of '0' because I assumed the missing values are due to an order made by clinets on the date of data collection (since it is not a full day since the last order, it likely didnt register a value for it)

In [39]:
# Step 7 Run a check for duplicte
df_dups= df_ords[df_ords.duplicated()]

In [40]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, hour_at_time_of_order, days_since_prior_order]
Index: []

In [41]:
# Step 8 Address the duplicates 

I didnt find any duplicate values

In [44]:
# Step 9 Export data as .csv
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'ords_checked.csv'), index=False)


In [45]:
df_ords.shape

(3421083, 7)